### ISSUE TITLE
Keras no hidden layer != simple linear regression ??

### ISSUE THREAD

**Issue by PanWu** [fchollet/keras#2576]
I am trying to run Keras with no hidden layer, and compare the result with a simple linear regression in sklearn. The data is generated with sklearn as below:

import matplotlib.pyplot as plt
%matplotlib inline
from sklearn import datasets, cross_validation
data_reg = datasets.make_regression(n_samples=10000, n_features=100, n_informative=100,
                                    n_targets=1, noise=100, random_state=0)

The sklearn code & result is:
from sklearn import linear_model
lm_lr = linear_model.LinearRegression()
lm_lr.fit(data_reg[0], data_reg[1])
plt.scatter(data_reg[1], lm_lr.predict(data_reg[0]))
![image](https://cloud.githubusercontent.com/assets/7094433/14945112/3114b752-0fbe-11e6-9560-34e426f9c3a3.png)

The Keras code & result is:
from keras.models import Sequential
from keras.layers import Input, Dense, Activation
model = Sequential()  
model.add(Dense(1, input_dim=100))
model.compile(loss="mse", optimizer="rmsprop")
model.fit(data_reg[0], data_reg[1], nb_epoch=10, batch_size=16)
plt.scatter(data_reg[1], model.predict(data_reg[0]))
![image](https://cloud.githubusercontent.com/assets/7094433/14945096/bad88ef6-0fbd-11e6-8638-72147a67be21.png)

It is obvious that Keras result is has the incorrect scale from -70 to 70, while the original Y scale is from -2000 to 2000. 

Could anyone help with understanding why Keras with no hidden layer does not recover the linear regression results?

**Comment by sschnug** [comment]
Just a few thoughts, as i'm no expert.

- Did you try plain SGD as optimizer? RMSprop looks quite complicated and maybe there is stuff happening which is hard to explain
- Even if using SGD: convergence is only guaranteed (theoretically) for a "good" learning rate; so try different rates
- SGD and all others optimizers (for convex functions) tend to like normalized data; you should preprocess
- Maybe you should add an additional input, playing the offset / bias (i'm not sure about that)

**Comment by the-moliver** [comment]
I think your main issue is the scale of your data. The optimizers in Keras are designed to take pretty small steps, so if the weight you need on a regressor is really big, it will take forever to get there. 
Rule of thumb: Z-score everything before feeding anything to a package like Keras.

**Comment by PanWu** [comment]
Thanks a lot! Right, it is about the scale of the Y data, once I normalize it to [-1, 1], the result looks quite similar. Also, once changing from rmsprop into SGD, the result is almost the same. 

scale = abs(data_reg[1]).max()
data_reg_new = data_reg[1] / scale

print 'no-hidden layer: ', metrics.mean_squared_error(data_reg_new, model.predict(data_reg[0]))
print 'linear regression: ', metrics.mean_squared_error(data_reg_new, lm_lr.predict(data_reg[0]))  

result is: 

no-hidden layer:  0.00202381332735
linear regression:  0.00201847959081

Thank you @the-moliver  and @sschnug for the help!

**Comment by PanWu** [comment]
Once extra thinking: 
    Would it be nice to have Keras automatic scale the data, and then re-scale it back for output/prediction? This could make Keras more convenient, without those manual scaling step.

**Comment by malzantot** [comment]
I don't agree about having Keras automically scaling the data, since in some situation the user wants to try training algorithm without scaling the data. However, I would suggest that it might be useful if we add a layer (Preprocessing Layer) that user might add to the model if he wants to normalize the features.

**Comment by sschnug** [comment]
+1 for NOT scaling per default.

And one more tiny note for non-toy projects: use scikit-learn's preprocessing tools for pleasant normalization (less error-prone when working with multiple data-transformations).

**Comment by PanWu** [comment]
This is good point, better to use sklearn to do the preprocessing & keep the flexibility of Keras. Thank you @malzantot, @sschnug  for the discussion!
